Option to create groups for std, external crates, and other imports

MattX opened the pull request:
This adds a configuration flag, `reorder_imports_opinionated`, that causes existing import groups to be discarded and replaced by three groups for standard imports, other imports, and `self`/`super`/`crate` imports.

The option name isn't great, we can change it if there are better ideas.

I know @calebcartwright raised a possible issue in https://github.com/rust-lang/rustfmt/issues/4107#issuecomment-609461946, but I'm not sure if it is still a blocker.

Resolves #4107.

calebcartwright commented:
Thanks for the PR @MattX! Excited to see this one. I don't think I'll have a chance to give this a review tonight, but a few initial thoughts for you:

* I think we'll want to use a more rich type for the config option instead of a simple boolean, for example an enum where each variant can represent a set of the groupings/order (even if there's only two variants in the beginning). I could see different camps wanting this feature albeit with different ordering and only having a boolean option would make that tricky
* We probably need to bikeshed on the name a bit. To me it's more of a `imports_groupings` or `group_imports` or `import_grouping_strategy` or something else, but I don't think we want to have the word `opinionated` in there
* Let's add some additional tests that incorporate this new option in conjunction with the various [other import-related options](https://rust-lang.github.io/rustfmt/?version=master&search=import) to make sure there's no unexpected/undesired behavior when using multiple options together

calebcartwright commented:
> I know @calebcartwright raised a possible issue in #4107 (comment), but I'm not sure if it is still a blocker.

The bug with comment association will be a blocker for being able to stabilize this option, but not a blocker for being able to make it available on nightly as an unstable option

MattX commented:
Thanks for the feedback!

- I've renamed the option to `group_imports` for now. I'm happy to change it to anything else.
- I've made the parameter into an enum. Not sure how to describe the current regrouping behavior, so I've called it `StdExternalCrate` for now. Also happy to change that.
- I've added two tests to check the interaction with `merge_imports` and `reorder_imports`.

The way I've implemented it, this new option has no effect when `reorder_imports` is `false`. This is more because it's more straightforward to implement than because it makes sense. After all, it seems possible to group imports without sorting them within each group (but I don't know if this is really something people want).

In any case, it might make sense to add a warning or error when `reorder_imports = false` and `group_imports != None`. I didn't see existing machinery to validate the config this way, so I haven't added this for now.

MattX commented on src/formatting/reorder.rs:
```diff
@@ -268,6 +282,38 @@ fn contains_macro_use_attr(attrs: &[ast::Attribute]) -> bool {
+    let mut local_imports = Vec::new();
+
+    for ut in uts.into_iter() {
+        if ut.path.is_empty() {
```
I'm assuming this should never happen? Same for having a Glob or List as the first element (line 306). I can panic here instead if this is true.

calebcartwright replied:
Unless it substantially impacts the implementation, I'd prefer to add the new feature without introducing panics so I think the check and continue here is fine :+1: 

There's only a couple explicit panics within rustfmt, but these are to address programmer errors (like a rustfmt developer calculating a bad span), and while there's a handful of `unreachable` cases, we do see users report panics from time to time. Better safe than sorry here :smile:

calebcartwright commented:
Thanks for the updates, I like the new name better. 

> The way I've implemented it, this new option has no effect when reorder_imports is false. This is more because it's more straightforward to implement than because it makes sense. After all, it seems possible to group imports without sorting them within each group (but I don't know if this is really something people want).

I was wondering about this one. I don't think we'll be able to couple the two options together and have this one only be operational one another has a certain value. Logically it makes sense that the majority of folks that would want an enforced grouping (and a sorting of those groups) would

calebcartwright commented on Configurations.md:
````diff
@@ -2008,6 +2008,56 @@ use dolor;
 use sit;
 ```
 
+## `group_imports`
+
+Discard existing import groups, and create three groups for:
+1. `std`, `core` and `alloc`,
+2. external crates,
+3. `self`, `super` and `crate` imports.
+
+Within each group, imports are sorted as with `reorder_imports`.
+
+This has no effect is `reorder_imports` is `false`.
+
+- **Default value**: `None`
+- **Possible values**: `None`, `StdExternalCrate`
+- **Stable**: No
+
+#### `None` (default):
````
Let's keep the description of the option generic about what the option does/what it controls as opposed to describing the behavior of a particular supported variant (e.g. `Controls the strategy for how imports are grouped together...`)

And given the denotations of `None` within Rust, I'd suggest we use `Preserve` instead, which will also be consistent with a few other rustfmt config option variants

MattX commented:
Thanks for the feedback! I was actually able to decouple ordering and grouping without much difficulty. I've also renamed `None` to `Preserve`.

calebcartwright commented on src/config.rs:
```diff
@@ -78,6 +78,8 @@ create_config! {
     imports_layout: ListTactic, ListTactic::Mixed, false, "Item layout inside a import block";
     merge_imports: bool, false, false, "Merge imports";
+    group_imports: GroupImportsTactic, GroupImportsTactic::Preserve, false,
+        "Reorganize import groups";
```
Minor nit but could we update this with something similar to what was changed in the Configurations.md file, like `Controls the strategy for how imports are grouped together`?

calebcartwright commented on src/formatting/reorder.rs:
```diff
@@ -227,19 +229,34 @@ fn rewrite_reorderable_items(
             if context.config.merge_imports() {
                 normalized_items = merge_use_trees(normalized_items);
             }
-            normalized_items.sort();
```
I'll confess this gave me pause at first glance, but see why we had to wrap the sorting behind the `reorder_imports` check here now since we can get here even if `reorder_imports` false with non-Preservation group import strategies.

calebcartwright reviewed (commented):
Excellent work @MattX, thank you! One super minor item left inline, but otherwise this is good to go from my perspective

calebcartwright commented on tests/source/group_imports-merge_imports.rs:
```diff
@@ -0,0 +1,17 @@
+// rustfmt-group_imports: StdExternalCrate
```
Oh and forgot to mention this earlier, but may be worth moving these tests to a new `group_imports` directory under `tests/{source/target}/configs/` where the majority of config option specific tests are housed. That should help simplify the file names and it'd probably also be worth using the name of the variant in the file name too.

calebcartwright reviewed (approved):
:+1:

calebcartwright commented:
Everything LGTM to me @MattX. Going to hold off a bit on merging though so @topecongiro can weigh in since this is a new config option

MattX commented:
Sounds good!

MattX commented:
Once it's reviewed, any chance this can be tagged as hacktoberfest-accepted? I want my t-shirt :p

topecongiro commented on src/formatting/reorder.rs:
```diff
@@ -227,19 +229,34 @@ fn rewrite_reorderable_items(
+                .collect::<Option<Vec<_>>>()?;
 
-            wrap_reorderable_items(context, &item_vec, nested_shape)
+            Some(item_vec.join("\n\n"))
```
We need to respect the current indentation here to indent inside inner modules correctly. E.g.,

```rust
mod test {
    use crate::foo::bar;
    use std::path;
}
```

MattX replied:
Good catch, I've added code to handle this case but I don't know if it's the best way to do things. Let me know if it's not.

topecongiro reviewed (changes requested):
Thank you for the PR! I think that the current implementation is good enough for the first round of implementing this feature, except for the inline comment. Would you mind fixing it and adding a test?

MattX commented:
Looks like the `rustc_ap_rustc_data_structures` crate is failing to compile in integ tests. The tests work for me locally (I'm probably using a different minor version?). I'll retry the checks when this is fixed.

calebcartwright commented:
There's a broken toolstate issue on nightly at the moment that requires us to bump the rustc-ap crates here. Those crates (snapshots of rustc internals) are published once a week, and I'm waiting to grab the next one which should be done in ~16 hours or so.

MattX commented:
Hey, just pinging on this to make sure it gets merged before it's too outdated. Let me know if there's anything that still needs to be done on my end.

calebcartwright commented:
@topecongiro going to go ahead and merge as the change you requested appears to have been resolved

calebcartwright commented:
@MattX - note that while this has been merged, and thus immediately available for anyone building rustfmt 2.0 from source, it's not available in the version of rustfmt distributed via rustup as that's on a different branch.

If you'd like to see this new feature supported in a released version of rustfmt, that will need to be backported to a target 1.x branch ([1.4.27](https://github.com/rust-lang/rustfmt/tree/rustfmt-1.4.27) at the moment). If you're interested in such a PR please let us know, otherwise I'll try to take a look at cherry-picking/backporting at some point down the road

MattX commented:
Ok! I'm opening a new PR for 1.4.27.
